src/Parsers/Access/ParserCreateUserQuery.cpp

```diff
@@ -107,6 +109,10 @@ namespace
                     return false;
 
                 value = ast->as<const ASTLiteral &>().value.safeGet<String>();
+                if (ParserStringLiteral{}.parse(pos, ast, expected))
+                {
+                    parsed_salt = ast->as<const ASTLiteral &>().value.safeGet<String>();
+                }
```

Comment:
I think this will allow to provide salt for `sha256_password`, and then screw random salt generation and appending it to password below:
```
CREATE USER user1 IDENTIFIED WITH sha256_password BY 'qwe123' 'ddd'
```

Also, maybe we need some syntax when providing salt - two string literals one after other looks weird, maybe like:
```
CREATE USER user1 IDENTIFIED WITH sha256_hash BY '<hash>' SALT '<salt>'
```

Reply:
There is no syntax change for CREATE USER command. to handle below two scenarios added this code.
1) During user creation using CREATE USER command:- Clickhouse will generate the salt and this code wont have any effect. UUID.sql file will be saved with ATTACH command ( salt will be added next to password).
2) Clickhouse server restart:- When CH server restarts, it will rebuild the user records by reading <UUID>.sql files under access directory. This code reads the salt for that particular user and will be used for authentication scenario.

Reply:
I think there is a bug though:
```
ClickHouse-ubuntu :) CREATE USER user1 IDENTIFIED WITH sha256_password BY 'qwe123' 'lll'

CREATE USER user1 IDENTIFIED WITH sha256_hash BY '18138372FAD4B94533CD4881F03DC6C69296DD897234E0CEE83F727E2E6B1F63' 'lll'

Query id: 7fff6d51-8192-4333-9681-8ec0009ac2e3

Ok.

0 rows in set. Elapsed: 0.002 sec.

ClickHouse-ubuntu :) select hex(SHA256('qwe123'))

SELECT hex(SHA256('qwe123'))

Query id: 33bb534f-1ea6-4995-8dda-9ef0021a309a

┌─hex(SHA256('qwe123'))────────────────────────────────────────────┐
│ 18138372FAD4B94533CD4881F03DC6C69296DD897234E0CEE83F727E2E6B1F63 │
└──────────────────────────────────────────────────────────────────┘

1 rows in set. Elapsed: 0.006 sec.
```

the bug is - you can provide salt for sha256_password - then salt will not be generated, but taken from CREATE USER, and never appended to password, hash will be generated from plain password too and stored in UUID.sql

Reply:
Hi , Thanks.
this feature is valid only for option sha256_password. as we need plain text password, which can be appended with generated salt before hashing.
for sha256_hash use has to provide password hash as input to the CREATE USER command.  We cannt covert back corresponding string for the hash given.

Reply:
I fixed this scenario. Thanks for finding it.

Reply:
> I think this will allow to provide salt for `sha256_password`, and then screw random salt generation and appending it to password below:
> 
> ```
> CREATE USER user1 IDENTIFIED WITH sha256_password BY 'qwe123' 'ddd'
> ```
> 
> Also, maybe we need some syntax when providing salt - two string literals one after other looks weird, maybe like:
> 
> ```
> CREATE USER user1 IDENTIFIED WITH sha256_hash BY '<hash>' SALT '<salt>'
> ```

keyword SALT has been added in the result to indicate salt value